crates/hc/src/calls.rs

```diff
@@ -0,0 +1,514 @@
+use holochain_types::prelude::DnaHash;
+use holochain_types::prelude::InstallAppDnaPayload;
+use holochain_types::prelude::InstallAppPayload;
+use holochain_types::prelude::InstalledCell;
+use portpicker::is_free;
+use portpicker::pick_unused_port;
+use std::convert::TryFrom;
+
+use crate::cmds::Existing;
+use crate::expect_match;
+use crate::ports::get_secondary_admin_ports;
+use crate::run::run_async;
+use crate::CmdRunner;
+use structopt::StructOpt;
+
+#[doc(hidden)]
+#[derive(Debug, StructOpt)]
+pub struct Call {
+    #[structopt(flatten)]
+    pub existing: Existing,
+    #[structopt(short, long, conflicts_with_all = &["existing_paths", "existing_indices"], value_delimiter = ",")]
+    /// Ports to running conductor admin interfaces.
+    /// If this is empty existing setups will be used.
+    /// Cannot be combined with existing setups.
+    pub running: Vec<u16>,
+    #[structopt(subcommand)]
+    /// The admin request you want to make.
+    pub call: AdminRequestCli,
+}
+
+// Docs have different use for structopt
+// so documenting everything doesn't make sense.
+#[allow(missing_docs)]
+#[derive(Debug, StructOpt, Clone)]
+pub enum AdminRequestCli {
+    AddAdminWs(AddAdminWs),
+    AddAppWs(AddAppWs),
+    InstallApp(InstallApp),
+    /// Calls AdminRequest::ListDnas.
+    ListDnas,
+    /// Calls AdminRequest::GenerateAgentPubKey.
+    NewAgent,
+    /// Calls AdminRequest::ListCellIds.
+    ListCells,
+    /// Calls AdminRequest::ListActiveApps.
+    ListActiveApps,
+    ActivateApp(ActivateApp),
+    DeactivateApp(DeactivateApp),
+    DumpState(DumpState),
+    /// Calls AdminRequest::AddAgentInfo.
+    /// [Unimplemented].
+    AddAgents,
+    ListAgents(ListAgents),
+}
+#[derive(Debug, StructOpt, Clone)]
+/// Calls AdminRequest::AddAdminInterfaces
+/// and adds another admin interface.
+pub struct AddAdminWs {
+    /// Optional port number.
+    /// Defaults to assigned by OS.
+    pub port: Option<u16>,
+}
+
+#[derive(Debug, StructOpt, Clone)]
+/// Calls AdminRequest::AttachAppInterface
+/// and adds another app interface.
+pub struct AddAppWs {
+    /// Optional port number.
+    /// Defaults to assigned by OS.
+    pub port: Option<u16>,
+}
+
+#[derive(Debug, StructOpt, Clone)]
+/// Calls AdminRequest::InstallApp
+/// and installs a new app.
+///
+/// Setting properties and membrane proofs is not
+/// yet supported.
+/// CellNicks are set to `my-app-0`, `my-app-1` etc.
+pub struct InstallApp {
+    #[structopt(short, long, default_value = "test-app")]
+    /// Sets the InstalledAppId.
+    pub app_id: String,
+    #[structopt(short, long, parse(try_from_str = parse_agent_key))]
+    /// If not set then a key will be generated.
+    /// Agent key is Base64 (same format that is used in logs).
+    /// e.g. `uhCAk71wNXTv7lstvi4PfUr_JDvxLucF9WzUgWPNIEZIoPGMF4b_o`
+    pub agent_key: Option<AgentPubKey>,
+    #[structopt(required = true, min_values = 1)]
+    /// List of dnas to install.
+    pub dnas: Vec<PathBuf>,
+}
```

Comment:
That's ok this is meant to reflect the current state of admin api. If it goes away then it will show up as compile errors here and we will need to fix it but that keeps everything in sync unlike javascript.